src/lib/ledger_catchup/super_catchup.ml

```diff
@@ -1485,15 +1472,16 @@ let%test_module "Ledger_catchup tests" =
         (Broadcast_pipe.Reader.peek Catchup_jobs.reader) ;
       [%log info] "target_best_tip_path length: %d"
         (List.length target_best_tip_path) ;
+      let target_best_tip_tree = Rose_tree.of_list_exn target_best_tip_path in
       [%log info] "breadcrumb_list length: %d" (List.length breadcrumb_list) ;
       let catchup_breadcrumbs_are_best_tip_path =
-        Rose_tree.equal (Rose_tree.of_list_exn target_best_tip_path)
-          breadcrumbs_tree ~f:(fun breadcrumb_tree1 breadcrumb_tree2 ->
-            External_transition.Validated.equal
-              (Transition_frontier.Breadcrumb.validated_transition
-                 breadcrumb_tree1)
-              (Transition_frontier.Breadcrumb.validated_transition
-                 breadcrumb_tree2))
+        Rose_tree.equal target_best_tip_tree breadcrumbs_tree ~f:(fun br1 br2 ->
+            let b1 = Transition_frontier.Breadcrumb.validated_transition br1 in
+            let b2 = Transition_frontier.Breadcrumb.validated_transition br2 in
+            (* We force evaluation of state body hash for both blocks for further equality check *)
+            let _hash1 = Mina_block.Validated.state_body_hash b1 in
```

Comment:
I'm not sure if I follow the necessity of this. If the state body hashes are equal, that implies that the state body hashes are equal, even if we don't have those values cached.

Reply:
State body hash is stored in a mutable variable which is set once the value is actually calculated.

Reply:
The problem arises when we compare this field for blocks computed in different code paths: one might have mutable variable set to `None` (because state body hash was not ever calculated up until the moment of execution of the line) and the other is `Some x`. Hence I force that both state hashes are evaluated and equal to `Some x`.